### PR TITLE
fix: Modify doubleEscape to handle path with extra backslashes and launch edge without browser key

### LINKED ISSUE #14716: Cypress --browser option not working on Windows for Edge browsers with full path
### Current behavior
Facing an issue while launching **Edge browsers** with the full path on **Windows 10** using `--browser`
I am running the following command:

```
cypress run --browser "C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe"
```

and the output is

```
We could not identify a known browser at the path you provided: `C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe`

The output from the command we ran was:

 Unable to find browser with path C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe
```

### Desired behavior
Should open edge browser from the given path just like it does for chrome and firefox.

### Test code to reproduce
Not dependent on test/specs, I was using the default example folder for this.

### Versions
Cypress Version: 6.3.0
Browser: MS Edge 88
OS: Windows 10

### Work-around
I Will try to raise a PR with a fix.

### LINKED ISSUE #14719: Cypress --browser crashes for paths with double backslashes (\\) on Windows (Chrome and Edge)
### Current behavior
While triggering the build, my ruby script is escaping the backslash making it `C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe`. For this path, Cypress launches the browser with an error and can't proceed with tests. Same behavior on MS Edge as well.

Note: Works fine for Firefox

#### Logs
```
C:~Path> cypress run --browser "C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe"


====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    6.3.0                                                                              │
  │ Browser:    Custom Chrome                                                                      │
  │ Specs:      19 found (examples\actions.spec.js, examples\aliasing.spec.js, examples\assertions │
  │             .spec.js, examples\connectors.spec.js, examples\cookies.spec.js, examples\cypress_ │
  │             api.spec.js, examples\files.spec.js, examples\local_storage.spec.js, examples\loca │
  │             tion.spec.j...)                                                                    │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────

  Running:  examples\actions.spec.js                                                       (1 of 19)

Timed out waiting for the browser to connect. Retrying...


Timed out waiting for the browser to connect. Retrying again...

The browser never connected. Something is wrong. The tests cannot run. Aborting...

The browser never connected. Something is wrong. The tests cannot run. Aborting...
```

#### Screenshot
image

Works fine when I remove extra `\` in the path
```
C:~Path> cypress run --browser "C:\Program Files (x86)\Google\Chrome\Application\chrome.exe"
```

### Desired behavior
Should launch browser correctly and run tests end to end for all win32 file paths. At lease handle double escapes :) 

### Test code to reproduce
Just running the default example with `--browser` and browser path with `\\`, you should get this on a Windows 10 machine very easily.

### Versions
Cypress: 6.3.0
OS:  Windows 10
Browser: Chrome 87, Edge 88 (Have tried lower version as well)

### PR DESCRIPTION
**Pull request by SagarGaniga**
- Closes #14719 
- Closes #14716

### User facing changelog
-  Fixes the issue with `--browser` for edge browsers on windows, allows users to pass just path to edge.exe and launch the browser.
- Fixes the issue with `--browser` to launch browsers when the path has extra backslashes on windows

### Additional details
- Why was this change necessary? **_To allow edge and chrome to launch on windows properly with --browser_**
- What is affected by this change? **_Browser detection in windows with --browser arg_**

`--browser` was breaking on windows for every browser but it was fixed in v5.4.0. But it still has issues 
- launching edge browsers when just the path is given without browser key
- launching edge and chrome when given path has multiple escaped \\
This PR adds code to handle edge paths and modifies `doubleEscape` function with the help of `path.win32` module which is capable of handling windows specific path separators in `windows/index.ts` 

This PR also adds some test cases to make sure these use cases are always tested.

### How has the user experience changed?
Users can pass edge and chrome paths with more freedom around escaped chars.
Users can now 
- Launch edge like ``` cypress run --browser "C:\\foo\\bar\\edge.exe" ```
- Launch chrome like ``` cypress run --browser "C:\\\\foo\\\\bar\\\\chrome.exe" ```

### PR Tasks
- [x] Have tests been added/updated?

### CODE REVIEW

**Comment by SagarGaniga** [comment]
Can someone please review this PR? 
`ci/circleci: server-e2e-tests-chrome` is failing repeatedly and I am not understanding what is the exact issue with it, can someone please check those errors?

**Comment by jennifer-shehane** [comment]
@SagarGaniga The server-e2e-tests-chrome failing are unrelated to this PR. A recent change in our stdout warnings is causing some flake in our internal snapshots. Our team is hoping to review this PR next sprint.

File: packages/launcher/lib/windows/index.ts [hunk 1]
```diff
@@ -203,15 +207,21 @@ export function getPathData (pathStr: string): PathData {
...
   }
 
+  path = doubleEscape(path)
+
   if (pathStr.indexOf('chrome.exe') > -1) {
     return { path, browserKey: 'chrome' }
   }
 
+  if (pathStr.indexOf('edge.exe') > -1) {
+    return { path, browserKey: 'edge' }
```
Comment (flotwig) [inline packages/launcher/lib/windows/index.ts:222]: Ha, I guess this does technically work for msedge.exe too

File: packages/launcher/test/unit/windows_spec.ts [hunk 2]
```diff
@@ -149,23 +159,74 @@ describe('windows browser detection', () => {
...
-      expect(res.path).to.eq('C:\\foo\\bar\\chrome.exe')
+    it('returns path and chrome given just nix path', () => {
+      const browserPath = 'C:/foo/bar/chrome.exe'
+      const res = windowsHelper.getPathData(browserPath)
+
+      expect(res.path).to.eq(doubleEscape(browserPath))
       expect(res.browserKey).to.eq('chrome')
     })
 
+    it('returns path and edge given just path', () => {
```
Comment (flotwig) [inline packages/launcher/test/unit/windows_spec.ts:201]: Could also use a test for msedge.exe - that is what #14716 is about

File: packages/launcher/test/unit/windows_spec.ts [hunk 3]
```diff
@@ -25,6 +25,10 @@ function stubBrowser (path: string, version: string) {
   .resolves(true)
 }
 
+function doubleEscape (s: string) {
+  return win32.join(...s.split(win32.sep)).replace(/\\/g, '\\\\')
+}
+
```
Comment (flotwig) [inline packages/launcher/test/unit/windows_spec.ts:31]: could you just import this from `windows.ts`?

**Review by flotwig — CHANGES_REQUESTED** [review]
Looks great! Just a few minor things

File: packages/launcher/lib/windows/index.ts (hunk 1 above)
Comment (SagarGaniga) [reply packages/launcher/lib/windows/index.ts:222]: Yeah, keeping that `browserKey` edge for msedge.exe too as we use `edge` everywhere else.

File: packages/launcher/test/unit/windows_spec.ts (hunk 2 above)
Comment (SagarGaniga) [reply packages/launcher/test/unit/windows_spec.ts]: Yes, updated

File: packages/launcher/test/unit/windows_spec.ts (hunk 3 above)
Comment (SagarGaniga) [reply packages/launcher/test/unit/windows_spec.ts]: I think you mean `packages/launcher/lib/windows/index.ts`

Exported that function, used it in the test case and added its own test cases

**Review by flotwig — APPROVED** [review]
Great stuff! Thanks for submitting this PR.

**Comment by flotwig** [comment]
While testing this PR, I noticed that the errors for path detection here are kinda abysmal... no fault of this PR tho

## this PR

image

image

image

image

## develop 

image

image

image

**Pull request merged**